filter/format_time.go

```diff
@@ -0,0 +1,194 @@
+- ` + "`RFC3339`" + ` format: "2006-01-02T15:04:05Z07:00"
+- ` + "`RFC3339Nano`" + ` format: "2006-01-02T15:04:05.999999999Z07:00"
+- ` + "`unix`" + ` unix epoch in seconds
+- ` + "`unixns`" + ` unix epoch in nanoseconds
+`
+	ansic       = "ANSIC"
+	unixdate    = "UnixDate"
+	rubydate    = "RubyDate"
+	rfc822      = "RFC822"
+	rfc822z     = "RFC822Z"
+	rfc850      = "RFC850"
+	rfc1123     = "RFC1123"
+	rfc1123z    = "RFC1123Z"
+	rfc3339     = "RFC3339"
+	rfc3339nano = "RFC3339Nano"
+	unix        = "unix"
+	unixns      = "unixns"
+)
+
+var FormatTimeDesc = baker.FilterDesc{
+	Name:   "FormatTime",
+	New:    NewFormatTime,
+	Config: &FormatTimeConfig{},
+	Help:   formatTimeHelp,
+}
+
+type FormatTimeConfig struct {
+	SrcField  string `help:"Field name of the input time" required:"true"`
+	DstField  string `help:"Field name of the output time" required:"true"`
+	SrcFormat string `help:"Format of the input time" required:"false" default:"UnixDate"`
+	DstFormat string `help:"Format of the output time" required:"false" default:"unix"`
+}
+
+func (cfg *FormatTimeConfig) fillDefaults() {
+	if cfg.SrcFormat == "" {
+		cfg.SrcFormat = unixdate
+	}
+	if cfg.DstFormat == "" {
+		cfg.DstFormat = unix
+	}
+}
+
+type FormatTime struct {
+	cfg *FormatTimeConfig
+
+	src    baker.FieldIndex
+	dst    baker.FieldIndex
+	parse  func(t []byte) (time.Time, error)
+	format func(t time.Time) []byte
+
+	// Shared state
+	numProcessedLines int64
+	numFilteredLines  int64
+}
+
+func NewFormatTime(cfg baker.FilterParams) (baker.Filter, error) {
+	dcfg := cfg.DecodedConfig.(*FormatTimeConfig)
+	dcfg.fillDefaults()
+
+	f := &FormatTime{cfg: dcfg}
+
+	idx, ok := cfg.FieldByName(dcfg.SrcField)
+	if !ok {
+		return nil, fmt.Errorf("unknown field %q", dcfg.SrcField)
+	}
+	f.src = idx
+
+	switch dcfg.SrcFormat {
+	case unix:
+		f.parse = func(b []byte) (time.Time, error) {
+			sec, err := strconv.ParseInt(string(b), 10, 64)
+			if err != nil {
+				return time.Time{}, err
+			}
+			return time.Unix(sec, 0), nil
+		}
+	case unixns:
+		f.parse = func(b []byte) (time.Time, error) {
+			nsec, err := strconv.ParseInt(string(b), 10, 64)
+			if err != nil {
+				return time.Time{}, err
+			}
+			return time.Unix(0, nsec), nil
+		}
+	default:
+		layout := formatToLayout(dcfg.SrcFormat)
+		f.parse = func(b []byte) (time.Time, error) {
+			t, err := time.Parse(layout, string(b))
+			if err != nil {
+				return time.Time{}, err
+			}
+			return t, nil
+		}
+	}
+
+	idx, ok = cfg.FieldByName(dcfg.DstField)
+	if !ok {
+		return nil, fmt.Errorf("unknown field %q", dcfg.DstField)
+	}
+	f.dst = idx
+
+	switch dcfg.DstFormat {
+	case unix:
+		f.format = func(t time.Time) []byte {
+			return strconv.AppendInt(nil, t.Unix(), 10)
+		}
+	case unixns:
+		f.format = func(t time.Time) []byte {
+			return strconv.AppendInt(nil, t.UnixNano(), 10)
+		}
+	default:
+		layout := formatToLayout(dcfg.DstFormat)
+		f.format = func(t time.Time) []byte {
+			return []byte(t.Format(layout))
+		}
+	}
+
+	return f, nil
```

Comment:
Before returning the instantiated filter, we should ensure that the user-provided formats are valid. I'm not sure all user-provided layout strings are valid ones so, since it's something we can know now, better to return an error and make baker fail early, rather than once the topology has started fetching records, etc, etc.

You can do that by simply using time.Now() with the `f.parse` and `f.format` closures. Also this should be present in unit-tests

Reply:
I already tried to find a way to check the time format of the user, but I think that it is not possible, see [playground](https://play.golang.org/p/MG7znIVcXLJ). The `time.Parse` function fails only if the layout and the time string do not match. The problem is that we cannot know the inputs of the user in advance.